pkg/lib/cockpit-connect-ssh.tsx

```diff
@@ -0,0 +1,604 @@
+                   onClose={onCancel}
+                   title={title}
+                   footer={<>
+                       <Button variant="primary" onClick={this.login} isLoading={this.state.in_progress}
+                               isDisabled={this.state.in_progress || (!offer_login_password && !offer_key_password) || !this.state.default_ssh_key || !this.props.error}>
+                           { submitText }
+                       </Button>
+                       <Button variant="link" className="btn-cancel" onClick={onCancel}>
+                           { _("Cancel") }
+                       </Button>
+                   </>}
+            >
+                <Stack hasGutter>
+                    { this.state.dialogError && <ModalError dialogError={this.state.dialogError} /> }
+                    {body}
+                </Stack>
+            </Modal>
+        );
+    }
+}
+
+const NotSupportedDialog = ({ host, error, dialogResult }: {
+    host: string,
+    error: cockpit.JsonObject,
+    dialogResult: DialogResult<void>,
+}) => (
+    <Modal id="ssh-not-supported-dialog" isOpen
+            position="top" variant="medium"
+            onClose={() => dialogResult.reject(error)}
+            title={_("Cockpit is not installed")}
+            footer={
+                <Button variant="link" className="btn-cancel" onClick={() => dialogResult.reject(error)}>
+                    { _("Close") }
+                </Button>
+            }
+    >
+        <Stack hasGutter>
+            <p>{cockpit.format(_("A compatible version of Cockpit is not installed on $0."), host)}</p>
+        </Stack>
+    </Modal>
+);
+
+const error_dialogs = {
+    "unknown-hostkey": UnknownHostDialog,
+    "invalid-hostkey": UnknownHostDialog,
+    "authentication-failed": ChangeAuthDialog,
+    "no-cockpit": NotSupportedDialog,
+};
+
+/**
+ * Set up SSH connection to a remote host
+ *
+ * Cockpit channels support running on a remote machine through SSH via the
+ * `host` channel option. This only works (without additional authentication
+ * options) if the SSH connection was already established (e.g. through the
+ * deprecated shell's "Add Host" feature), or can be established
+ * noninteractively (e.g. if you have a passwordless SSH key, or a special
+ * noninteractive configuration block in your ~/.ssh/config for the target host).
+ *
+ * For all other cases, call this function first. It shows various dialogs where
+ * the user can specify a login password or unlock their SSH key. If the user
+ * does not already have an SSH key, the dialog also offers to create one. It
+ * also offers to authorize the user's SSH key to the remote machine/user.
+ *
+ * Arguments:
+ * @dialog_context: The page's `DialogsContext`, see ./dialogs.tsx
+ * @host: Same `[user@]host[:port]` format as the channel option; must be
+ *        *exactly* the same as for opening the channel afterwards
+ * @user: Same as the channel option; overrides `user@` portion of @host
```

Comment:
So this is a wart of our channel optrions? One can say `{ host: "bob@barn.com", user: "bill" }` on the first open message, and then using `{ host: "bob@barn.com" }` on subsequent messages will actually use the SSH connection for bill? I find this confusing, but probably nothing new with this API.

Reply:
Should we maybe simply not document the possibility of adding "user@" to `host`? I think this might never have been intended, and only works because SSH allows it.

Reply:
Nah, let's leave it as it is, warts and all.